test/test_performance_specializations.jl

```diff
@@ -0,0 +1,50 @@
+module TestPerformanceSpecializations
+
+using Test
+using Trixi
+
+include("test_trixi.jl")
+
+# Start with a clean environment: remove Trixi output directory if it exists
+outdir = "out"
+isdir(outdir) && rm(outdir, recursive=true)
+
+
+@testset "Performance specializations" begin
+  @timed_testset "TreeMesh3D, flux_shima_etal" begin
+    trixi_include(@__MODULE__,
+      joinpath(examples_dir(), "tree_3d_dgsem", "elixir_euler_ec.jl"),
+      initial_refinement_level=1, tspan=(0.0, 0.0),
+      volume_flux=flux_shima_etal, surface_flux=flux_shima_etal)
+    u_ode = copy(sol.u[end])
+    du_ode = zero(u_ode)
+    u = Trixi.wrap_array(u_ode, semi)
+    du = Trixi.wrap_array(du_ode, semi)
+    nonconservative_terms = Trixi.have_nonconservative_terms(semi.equations)
+
+    du .= 0
+    Trixi.split_form_kernel!(
+      du, u, 1, semi.mesh,
+      nonconservative_terms, semi.equations,
+      semi.solver.volume_integral.volume_flux, semi.solver, semi.cache)
+    du_specialized = du[:, :, :, :, 1]
+
+    du .= 0
+    invoke(Trixi.split_form_kernel!,
```

Comment:
To get the non-optimized baseline version:
```julia
shell> git diff
diff --git a/src/solvers/dgsem_tree/dg_3d.jl b/src/solvers/dgsem_tree/dg_3d.jl
index dbee50b6..71f5cf78 100644
--- a/src/solvers/dgsem_tree/dg_3d.jl
+++ b/src/solvers/dgsem_tree/dg_3d.jl
@@ -242,6 +242,7 @@ end
   # true * [some floating point value] == [exactly the same floating point value]
   # This can (hopefully) be optimized away due to constant propagation.
   @unpack derivative_split = dg.basis
+  @info "split_form_kernel! plain"
 
   # Calculate volume integral in one element
   for k in eachnode(dg), j in eachnode(dg), i in eachnode(dg)
diff --git a/src/solvers/dgsem_tree/dg_3d_compressible_euler.jl b/src/solvers/dgsem_tree/dg_3d_compressible_euler.jl
index ef331030..8c07c423 100644
--- a/src/solvers/dgsem_tree/dg_3d_compressible_euler.jl
+++ b/src/solvers/dgsem_tree/dg_3d_compressible_euler.jl
@@ -27,6 +27,7 @@
   # true * [some floating point value] == [exactly the same floating point value]
   # This can (hopefully) be optimized away due to constant propagation.
   @unpack derivative_split = dg.basis
+  @info "split_form_kernel! optimized"
 
   # Create a temporary array that will be used to store the RHS with permuted
   # indices `[i, j, k, v]` to allow using SIMD instructions.

julia> include("test/test_performance_specializations.jl")
[...]
[ Info: split_form_kernel! optimized # from trixi_include
[...]
[ Info: split_form_kernel! optimized
[ Info: split_form_kernel! plain
[ Info: Testset TreeMesh3D, flux_shima_etal finished in 41.78921711 seconds.
Test Summary:               | Pass  Total
Performance specializations |    1      1
Main.TestPerformanceSpecializations
```
I added some comments in 8dd7c003.

Reply:
Perfect, thanks